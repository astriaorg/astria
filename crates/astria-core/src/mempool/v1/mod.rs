pub mod transaction_status;
