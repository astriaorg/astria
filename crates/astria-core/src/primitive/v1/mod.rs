pub mod asset;
pub mod u128;

use base64::{
    display::Base64Display,
    prelude::BASE64_STANDARD,
};
use sha2::{
    Digest as _,
    Sha256,
};

use crate::{
    generated::primitive::v1 as raw,
    Protobuf,
};

pub const ADDRESS_LEN: usize = 20;
pub const ROLLUP_ID_LEN: usize = 32;
pub const FEE_ASSET_ID_LEN: usize = 32;

impl Protobuf for merkle::Proof {
    type Error = merkle::audit::InvalidProof;
    type Raw = raw::Proof;

    fn try_from_raw_ref(raw: &Self::Raw) -> Result<Self, Self::Error> {
        // XXX: Implementing this by cloning is ok because `audit_path`
        //      has to be cloned always due to `UncheckedProof`'s constructor.
        Self::try_from_raw(raw.clone())
    }

    fn try_from_raw(raw: Self::Raw) -> Result<Self, Self::Error> {
        let Self::Raw {
            audit_path,
            leaf_index,
            tree_size,
        } = raw;
        let leaf_index = leaf_index.try_into().expect(
            "running on a machine with at least 64 bit pointer width and can convert from u64 to \
             usize",
        );
        let tree_size = tree_size.try_into().expect(
            "running on a machine with at least 64 bit pointer width and can convert from u64 to \
             usize",
        );
        Self::unchecked()
            .audit_path(audit_path.to_vec())
            .leaf_index(leaf_index)
            .tree_size(tree_size)
            .try_into_proof()
    }

    fn to_raw(&self) -> Self::Raw {
        // XXX: Implementing in terms of clone is ok because the fields would need to be cloned
        // anyway.
        self.clone().into_raw()
    }

    fn into_raw(self) -> Self::Raw {
        let merkle::audit::UncheckedProof {
            audit_path,
            leaf_index,
            tree_size,
        } = self.into_unchecked();
        Self::Raw {
            audit_path: audit_path.into(),
            leaf_index: leaf_index.try_into().expect(
                "running on a machine with at most 64 bit pointer width and can convert from \
                 usize to u64",
            ),
            tree_size: tree_size.try_into().expect(
                "running on a machine with at most 64 bit pointer width and can convert from \
                 usize to u64",
            ),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[cfg_attr(feature = "serde", serde(transparent))]
pub struct RollupId {
    #[cfg_attr(feature = "serde", serde(serialize_with = "crate::serde::base64"))]
    inner: [u8; 32],
}

impl RollupId {
    /// Creates a new rollup ID from a 32 byte array.
    ///
    /// Use this if you already have a 32 byte array. Prefer
    /// [`RollupId::from_unhashed_bytes`] if you have a clear text
    /// name what you want to use to identify your rollup.
    ///
    /// # Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// let bytes = [42u8; 32];
    /// let rollup_id = RollupId::new(bytes);
    /// assert_eq!(bytes, rollup_id.get());
    /// ```
    #[must_use]
    pub const fn new(inner: [u8; ROLLUP_ID_LEN]) -> Self {
        Self {
            inner,
        }
    }

    /// Returns the 32 bytes array representing the rollup ID.
    ///
    /// # Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// let bytes = [42u8; 32];
    /// let rollup_id = RollupId::new(bytes);
    /// assert_eq!(bytes, rollup_id.get());
    /// ```
    #[must_use]
    pub const fn get(self) -> [u8; 32] {
        self.inner
    }

    /// Creates a new rollup ID by applying Sha256 to `bytes`.
    ///
    /// Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// use sha2::{
    ///     Digest,
    ///     Sha256,
    /// };
    /// let name = "MyRollup-1";
    /// let hashed = Sha256::digest(name);
    /// let rollup_id = RollupId::from_unhashed_bytes(name);
    /// assert_eq!(rollup_id, RollupId::new(hashed.into()));
    /// ```
    #[must_use]
    pub fn from_unhashed_bytes<T: AsRef<[u8]>>(bytes: T) -> Self {
        Self {
            inner: Sha256::digest(bytes).into(),
        }
    }

    /// Allocates a vector from the fixed size array holding the rollup ID.
    ///
    /// # Examples
    /// ```
    /// use astria_core::primitive::v1::RollupId;
    /// let rollup_id = RollupId::new([42u8; 32]);
    /// assert_eq!(vec![42u8; 32], rollup_id.to_vec());
    /// ```
    #[must_use]
    pub fn to_vec(self) -> Vec<u8> {
        self.inner.to_vec()
    }

    /// Convert a byte slice to a rollup ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the byte slice was not 32 bytes long.
    pub fn try_from_slice(bytes: &[u8]) -> Result<Self, IncorrectRollupIdLength> {
        let inner =
            <[u8; ROLLUP_ID_LEN]>::try_from(bytes).map_err(|_| IncorrectRollupIdLength {
                received: bytes.len(),
            })?;
        Ok(Self::new(inner))
    }

    /// Converts a byte vector to a rollup ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the byte slice was not 32 bytes long.
    pub fn try_from_vec(bytes: Vec<u8>) -> Result<Self, IncorrectRollupIdLength> {
        let inner =
            <[u8; ROLLUP_ID_LEN]>::try_from(bytes).map_err(|bytes| IncorrectRollupIdLength {
                received: bytes.len(),
            })?;
        Ok(Self::new(inner))
    }

    #[must_use]
    pub fn to_raw(&self) -> raw::RollupId {
        raw::RollupId {
            inner: self.to_vec().into(),
        }
    }

    #[must_use]
    pub fn into_raw(self) -> raw::RollupId {
        raw::RollupId {
            inner: self.to_vec().into(),
        }
    }

    /// Converts from protobuf type to rust type for a rollup ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the byte slice was not 32 bytes long.
    pub fn try_from_raw(raw: &raw::RollupId) -> Result<Self, IncorrectRollupIdLength> {
        Self::try_from_slice(&raw.inner)
    }
}

impl AsRef<[u8]> for RollupId {
    fn as_ref(&self) -> &[u8] {
        &self.inner
    }
}

impl From<[u8; ROLLUP_ID_LEN]> for RollupId {
    fn from(inner: [u8; ROLLUP_ID_LEN]) -> Self {
        Self {
            inner,
        }
    }
}

impl From<&[u8; ROLLUP_ID_LEN]> for RollupId {
    fn from(inner: &[u8; ROLLUP_ID_LEN]) -> Self {
        Self {
            inner: *inner,
        }
    }
}

impl From<&RollupId> for RollupId {
    fn from(value: &RollupId) -> Self {
        *value
    }
}

impl std::fmt::Display for RollupId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Base64Display::new(self.as_ref(), &BASE64_STANDARD).fmt(f)
    }
}

#[derive(Debug, thiserror::Error)]
#[error("expected 32 bytes, got {received}")]
pub struct IncorrectRollupIdLength {
    received: usize,
}

/// Indicates that the protobuf response contained an array field that was not 20 bytes long.
#[derive(Debug, thiserror::Error)]
#[error("expected 20 bytes, got {received}")]
pub struct IncorrectAddressLength {
    received: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct Address(
    #[cfg_attr(feature = "serde", serde(serialize_with = "crate::serde::base64"))]
    [u8; ADDRESS_LEN],
);

impl Address {
    #[must_use]
    pub fn get(self) -> [u8; ADDRESS_LEN] {
        self.0
    }

    #[must_use]
    pub fn to_vec(&self) -> Vec<u8> {
        self.0.to_vec()
    }

    /// Construct a sequencer address from a [`ed25519_consensus::VerificationKey`].
    ///
    /// The first 20 bytes of the sha256 hash of the verification key is the address.
    #[must_use]
    // Silence the clippy lint because the function body asserts that the panic
    // cannot happen.
    #[allow(clippy::missing_panics_doc)]
    pub fn from_verification_key(public_key: ed25519_consensus::VerificationKey) -> Self {
        /// this ensures that `ADDRESS_LEN` is never accidentally changed to a value
        /// that would violate this assumption.
        #[allow(clippy::assertions_on_constants)]
        const _: () = assert!(ADDRESS_LEN <= 32);
        let bytes: [u8; 32] = Sha256::digest(public_key).into();
        Self::try_from_slice(&bytes[..ADDRESS_LEN])
            .expect("can convert 32 byte hash to 20 byte array")
    }

    /// Convert a byte slice to an address.
    ///
    /// # Errors
    ///
    /// Returns an error if the account buffer was not 20 bytes long.
    pub fn try_from_slice(bytes: &[u8]) -> Result<Self, IncorrectAddressLength> {
        let inner = <[u8; ADDRESS_LEN]>::try_from(bytes).map_err(|_| IncorrectAddressLength {
            received: bytes.len(),
        })?;
        Ok(Self::from_array(inner))
    }

    #[must_use]
    pub const fn from_array(array: [u8; ADDRESS_LEN]) -> Self {
        Self(array)
    }

    #[must_use]
    pub fn to_raw(&self) -> raw::Address {
        raw::Address {
            inner: self.to_vec().into(),
        }
    }

    #[must_use]
    pub fn into_raw(self) -> raw::Address {
        raw::Address {
            inner: self.to_vec().into(),
        }
    }

    /// Convert from protobuf to rust type an address.
    ///
    /// # Errors
    ///
    /// Returns an error if the account buffer was not 20 bytes long.
    pub fn try_from_raw(raw: &raw::Address) -> Result<Self, IncorrectAddressLength> {
        Self::try_from_slice(&raw.inner)
    }
}

impl AsRef<[u8]> for Address {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}

impl From<[u8; ADDRESS_LEN]> for Address {
    fn from(inner: [u8; ADDRESS_LEN]) -> Self {
        Self(inner)
    }
}

impl std::fmt::Display for Address {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Base64Display::new(self.as_ref(), &BASE64_STANDARD).fmt(f)
    }
}

/// Derive a [`merkle::Tree`] from an iterable.
///
/// It is the responsibility of the caller to ensure that the iterable is
/// deterministic. Prefer types like `Vec`, `BTreeMap` or `IndexMap` over
/// `HashMap`.
pub fn derive_merkle_tree_from_rollup_txs<'a, T: 'a, U: 'a>(rollup_ids_to_txs: T) -> merkle::Tree
where
    T: IntoIterator<Item = (&'a RollupId, &'a U)>,
    U: AsRef<[Vec<u8>]> + 'a + ?Sized,
{
    let mut tree = merkle::Tree::new();
    for (rollup_id, txs) in rollup_ids_to_txs {
        let root = merkle::Tree::from_leaves(txs.as_ref()).root();
        tree.build_leaf().write(rollup_id.as_ref()).write(&root);
    }
    tree
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;

    use super::{
        Address,
        IncorrectAddressLength,
    };

    #[test]
    fn account_of_20_bytes_is_converted_correctly() {
        let expected = Address([42; 20]);
        let account_vec = expected.0.to_vec();
        let actual = Address::try_from_slice(&account_vec).unwrap();
        assert_eq!(expected, actual);
    }

    #[track_caller]
    fn account_conversion_check(bad_account: &[u8]) {
        let error = Address::try_from_slice(bad_account);
        assert!(
            matches!(error, Err(IncorrectAddressLength { .. })),
            "converting form incorrect sized account succeeded where it should have failed"
        );
    }

    #[test]
    fn account_of_incorrect_length_gives_error() {
        account_conversion_check(&[42; 0]);
        account_conversion_check(&[42; 19]);
        account_conversion_check(&[42; 21]);
        account_conversion_check(&[42; 100]);
    }

    #[test]
    fn snapshots() {
        let address = Address([42; 20]);
        assert_json_snapshot!(address);
    }
}
