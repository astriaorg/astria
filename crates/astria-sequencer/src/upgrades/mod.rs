mod state_ext;
pub(crate) mod storage;
mod upgrades_handler;

pub(crate) use state_ext::{
    StateReadExt,
    StateWriteExt,
};
pub(crate) use upgrades_handler::UpgradesHandler;
