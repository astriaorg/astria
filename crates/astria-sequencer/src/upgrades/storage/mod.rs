pub(super) mod keys;
mod values;

pub(super) use values::ChangeInfo;
pub(crate) use values::Value;
