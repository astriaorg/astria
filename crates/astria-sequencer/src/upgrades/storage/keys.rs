use astria_core::upgrades::v1::{
    ChangeName,
    UpgradeName,
};

pub(in crate::upgrades) fn change(upgrade_name: &UpgradeName, change_name: &ChangeName) -> String {
    format!("upgrades/{upgrade_name}/{change_name}")
}

#[cfg(test)]
mod tests {
    use super::*;

    const COMPONENT_PREFIX: &str = "upgrades/";

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(change(
            &UpgradeName::new("aspen"),
            &ChangeName::new("change_1")
        ));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(
            change(&UpgradeName::new("aspen"), &ChangeName::new("change_1"))
                .starts_with(COMPONENT_PREFIX)
        );
    }
}
