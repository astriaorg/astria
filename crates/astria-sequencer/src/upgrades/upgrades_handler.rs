use std::{
    path::Path,
    time::Duration,
};

use astria_core::upgrades::v1::{
    ChangeHash,
    Upgrade,
    Upgrades,
};
use astria_eyre::{
    anyhow_to_eyre,
    eyre::{
        self,
        bail,
        eyre,
        OptionExt as _,
        Result,
        WrapErr as _,
    },
};
use cnidarium::{
    Snapshot,
    StateWrite,
};
use isahc::AsyncReadResponseExt;
use serde_json::Value;
use tendermint::consensus::params::VersionParams;
use tracing::{
    error,
    info,
    warn,
};

use super::{
    StateReadExt as _,
    StateWriteExt as _,
};
use crate::{
    app::{
        ShouldShutDown,
        StateReadExt as _,
        StateWriteExt,
    },
    oracles::price_feed::{
        market_map::component::MarketMapComponent,
        oracle::component::OracleComponent,
    },
};

pub(crate) struct UpgradesHandler {
    upgrades: Upgrades,
    cometbft_rpc_addr: String,
}

impl UpgradesHandler {
    pub(crate) fn new<P: AsRef<Path>>(
        upgrades_filepath: P,
        cometbft_rpc_addr: String,
    ) -> Result<Self> {
        let upgrades = Upgrades::new(upgrades_filepath).wrap_err("failed constructing upgrades")?;
        Ok(Self {
            upgrades,
            cometbft_rpc_addr,
        })
    }

    pub(crate) fn upgrades(&self) -> &Upgrades {
        &self.upgrades
    }

    /// Verifies that all historical upgrades have been applied.
    ///
    /// Returns an error if any has not been applied.
    pub(crate) async fn ensure_historical_upgrades_applied(
        &self,
        snapshot: &Snapshot,
    ) -> Result<()> {
        let next_block_height = next_block_height(snapshot).await?;

        for upgrade in self.upgrades.iter() {
            // The upgrades are ordered from lowest activation height to highest, so once we find
            // an upgrade with an activation height >= `next_block_height`, we can stop iterating
            // as all subsequent upgrades have activation heights > `next_block_height`.  In the
            // case where activation height == `next_block_height`, the upgrade is not a historical
            // one and is not expected to have been applied at the point of calling this function.
            if upgrade.activation_height() >= next_block_height {
                return Ok(());
            }

            let upgrade_name = upgrade.name();
            for change in upgrade.changes() {
                let change_name = change.name();
                let Some(stored_change_info) = snapshot
                    .get_upgrade_change_info(&upgrade_name, &change_name)
                    .await
                    .wrap_err("failed to get upgrade change hash")?
                else {
                    bail!(
                        "historical upgrade change `{upgrade_name}/{change_name}` has not been \
                         applied (wrong upgrade.json file provided?)",
                    );
                };
                let actual_info = change.info();
                if actual_info != stored_change_info {
                    bail!(
                        "upgrade change `{actual_info}` does not match stored info \
                         `{stored_change_info}` for `{upgrade_name}/{change_name}`",
                    );
                }
            }
        }
        Ok(())
    }

    /// Returns `ShouldShutDown::ShutDownForUpgrade` if any scheduled upgrade is due to be applied
    /// during the next block and this binary is not hard-coded to apply that upgrade.  Otherwise,
    /// returns `ShouldShutDown::ContinueRunning`.
    pub(crate) async fn should_shut_down(&self, snapshot: &Snapshot) -> Result<ShouldShutDown> {
        let next_block_height = next_block_height(snapshot).await?;

        for upgrade in self.upgrades.iter() {
            // The upgrades are ordered from lowest activation height to highest, so once we find an
            // upgrade with an activation height > `next_block_height`, we can stop iterating as all
            // subsequent upgrades have activation heights > `next_block_height`.
            if upgrade.activation_height() < next_block_height {
                continue;
            }
            if upgrade.activation_height() > next_block_height {
                return Ok(ShouldShutDown::ContinueRunning);
            }

            if !upgrade.shutdown_required() {
                continue;
            }

            let block_time = snapshot
                .get_block_timestamp()
                .await
                .wrap_err("failed getting latest block time from snapshot")?;
            let app_hash = snapshot
                .root_hash()
                .await
                .map_err(anyhow_to_eyre)
                .wrap_err("failed to get current root hash from snapshot")?;
            let hex_encoded_app_hash = hex::encode(app_hash.0);

            return Ok(ShouldShutDown::ShutDownForUpgrade {
                upgrade_activation_height: upgrade.activation_height(),
                block_time,
                hex_encoded_app_hash,
            });
        }

        Ok(ShouldShutDown::ContinueRunning)
    }

    /// Execute any changes to global state required as part of any upgrade with an activation
    /// height == `block_height`.
    ///
    /// At a minimum, the `info` of each `Change` in such an upgrade must be written to verifiable
    /// storage.
    ///
    /// Returns an empty `Vec` if no upgrade was executed.
    pub(crate) fn execute_upgrade_if_due<S: StateWrite>(
        &mut self,
        mut state: S,
        block_height: tendermint::block::Height,
    ) -> Result<Vec<ChangeHash>> {
        let Some(upgrade) = self
            .upgrades
            .upgrade_activating_at_height(block_height.value())
        else {
            return Ok(vec![]);
        };
        let upgrade_name = upgrade.name();
        let mut change_hashes = vec![];
        for change in upgrade.changes() {
            change_hashes.push(change.calculate_hash());
            state
                .put_upgrade_change_info(&upgrade_name, change)
                .wrap_err("failed to put upgrade change info")?;
            info!(upgrade = %upgrade_name, change = %change.name(), "executed upgrade change");
        }

        // NOTE: any further state changes specific to individual upgrades should be
        //       executed here after matching on the upgrade variant.

        #[expect(
            irrefutable_let_patterns,
            reason = "will become refutable once we have more than one upgrade variant"
        )]
        if let Upgrade::Upgrade1(upgrade_1) = upgrade {
            let genesis_state = upgrade_1.price_feed_change().genesis();
            MarketMapComponent::handle_genesis(&mut state, genesis_state.market_map())
                .wrap_err("failed to handle market map genesis")?;
            info!("handled market map genesis");
            OracleComponent::handle_genesis(state, genesis_state.oracle())
                .wrap_err("failed to handle oracle genesis")?;
            info!("handled oracle genesis");
        }

        Ok(change_hashes)
    }

    /// Updates CometBFT consensus params as required by any upgrade with an activation height ==
    /// `block_height`.
    ///
    /// Returns `None` if no upgrade is due.
    ///
    /// If an upgrade is due, at a minimum, the ABCI application version should be increased.
    pub(crate) async fn end_block<S: StateWriteExt>(
        &self,
        mut state: S,
        block_height: tendermint::block::Height,
    ) -> Result<Option<tendermint::consensus::Params>> {
        let Some(upgrade) = self
            .upgrades
            .upgrade_activating_at_height(block_height.value())
        else {
            return Ok(None);
        };

        let mut params = match state
            .get_consensus_params()
            .await
            .wrap_err("failed to get consensus params from storage")?
        {
            Some(value) => value,
            None => self
                .get_consensus_params_from_cometbft(block_height.value())
                .await
                .wrap_err("failed to get consensus params from cometbft")?,
        };

        let new_app_version = upgrade.app_version();
        if let Some(existing_app_version) = &params.version {
            if new_app_version <= existing_app_version.app {
                error!(
                    "new app version {new_app_version} should be greater than existing version {}",
                    existing_app_version.app
                );
            }
        }
        params.version = Some(VersionParams {
            app: new_app_version,
        });

        // NOTE: any further changes specific to individual upgrades should be applied here after
        //       matching on the upgrade variant.

        #[expect(
            irrefutable_let_patterns,
            reason = "will become refutable once we have more than one upgrade variant"
        )]
        if let Upgrade::Upgrade1(_) = upgrade {
            set_vote_extensions_enable_height_to_next_block_height(block_height, &mut params);
        }

        state
            .put_consensus_params(params.clone())
            .wrap_err("failed to put consensus params to storage")?;

        Ok(Some(params))
    }

    /// Returns the CometBFT consensus params by querying the given CometBFT endpoint.
    ///
    /// We need to specify the current block height as a query arg since otherwise CometBFT tries to
    /// use its view of current block, and since `FinalizeBlock` has not yet been called, this
    /// results in an error response.
    async fn get_consensus_params_from_cometbft(
        &self,
        block_height: u64,
    ) -> Result<tendermint::consensus::Params> {
        if cfg!(test) {
            bail!(
                "cannot query cometbft in tests; consensus params should be available to `App` as \
                 they are provided in `App::init_chain`"
            );
        }

        let uri = format!(
            "{}/consensus_params?height={block_height}",
            self.cometbft_rpc_addr
        );

        let max_retries = 66;
        let retry_config = tryhard::RetryFutureConfig::new(max_retries)
            .exponential_backoff(Duration::from_millis(10))
            .max_delay(Duration::from_secs(1))
            .on_retry(
                |attempt, next_delay: Option<Duration>, error: &eyre::Report| {
                    let wait_duration = next_delay
                        .map(telemetry::display::format_duration)
                        .map(tracing::field::display);
                    warn!(
                        error = error.as_ref() as &dyn std::error::Error,
                        attempt,
                        wait_duration,
                        "failed to get consensus params from cometbft; retrying after backoff",
                    );
                    async {}
                },
            );

        tryhard::retry_fn(|| try_get_consensus_params_from_cometbft(&uri))
            .with_config(retry_config)
            .await
            .wrap_err_with(|| {
                format!(
                    "failed to get consensus params from {uri} after {max_retries} retries; \
                     giving up"
                )
            })
    }
}

#[cfg(any(test, feature = "benchmark"))]
impl From<Upgrades> for UpgradesHandler {
    fn from(upgrades: Upgrades) -> Self {
        Self {
            upgrades,
            cometbft_rpc_addr: String::new(),
        }
    }
}

async fn next_block_height(snapshot: &Snapshot) -> Result<u64> {
    snapshot
        .get_block_height()
        .await
        .unwrap_or_default()
        .checked_add(1)
        .ok_or_eyre("overflowed getting next block height")
}

fn set_vote_extensions_enable_height_to_next_block_height(
    current_block_height: tendermint::block::Height,
    consensus_params: &mut tendermint::consensus::Params,
) {
    // Set the vote_extensions_enable_height as the next block height (it must be a future
    // height to be valid).
    let new_enable_height = current_block_height.increment();
    if let Some(existing_enable_height) = consensus_params.abci.vote_extensions_enable_height {
        // If vote extensions are already enabled, they cannot be disabled, and the
        // `vote_extensions_enable_height` cannot be changed.
        if existing_enable_height.value() != 0 {
            error!(
                %existing_enable_height, %new_enable_height,
                "vote extensions enable height already set; ignoring update",
            );
            return;
        }
    }
    consensus_params.abci.vote_extensions_enable_height = Some(new_enable_height);
}

async fn try_get_consensus_params_from_cometbft(
    uri: &str,
) -> Result<tendermint::consensus::Params> {
    let blocking_getter = async {
        isahc::get_async(uri)
            .await
            .wrap_err("failed to get consensus params")?
            .text()
            .await
            .wrap_err("failed to parse consensus params response as UTF-8 string")
    };

    let response = tokio::time::timeout(Duration::from_secs(1), blocking_getter)
        .await
        .wrap_err("timed out fetching consensus params")??
        .trim()
        .to_string();
    let json_rpc_response: Value = serde_json::from_str(&response).wrap_err_with(|| {
        format!("failed to parse consensus params response `{response}` as json")
    })?;
    let json_params = json_rpc_response
        .get("result")
        .and_then(|result| result.get("consensus_params"))
        .ok_or_else(|| eyre!("missing `result` in consensus params response `{response}`"))?
        .clone();
    serde_json::from_value::<tendermint::consensus::Params>(json_params).wrap_err_with(|| {
        format!(
            "failed to parse `result.consensus_params` as consensus params in response \
             `{response}`"
        )
    })
}
