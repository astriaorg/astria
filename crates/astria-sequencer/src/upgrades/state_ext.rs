use astria_core::upgrades::v1::{
    Change,
    ChangeInfo,
    ChangeName,
    UpgradeName,
};
use astria_eyre::{
    anyhow_to_eyre,
    eyre::{
        Result,
        WrapErr as _,
    },
};
use async_trait::async_trait;
use cnidarium::{
    StateRead,
    StateWrite,
};
use tracing::instrument;

use super::storage::{
    self,
    keys,
};
use crate::storage::StoredValue;

#[async_trait]
pub(crate) trait StateReadExt: StateRead {
    #[instrument(skip_all)]
    async fn get_upgrade_change_info(
        &self,
        upgrade_name: &UpgradeName,
        change_name: &ChangeName,
    ) -> Result<Option<ChangeInfo>> {
        let Some(bytes) = self
            .get_raw(&keys::change(upgrade_name, change_name))
            .await
            .map_err(anyhow_to_eyre)
            .wrap_err("failed reading change info from state")?
        else {
            return Ok(None);
        };
        StoredValue::deserialize(&bytes)
            .and_then(|value| {
                storage::ChangeInfo::try_from(value).map(|info| Some(ChangeInfo::from(info)))
            })
            .wrap_err("invalid change info bytes")
    }
}

impl<T: StateRead> StateReadExt for T {}

#[async_trait]
pub(crate) trait StateWriteExt: StateWrite {
    #[instrument(skip_all)]
    fn put_upgrade_change_info(
        &mut self,
        upgrade_name: &UpgradeName,
        change: &dyn Change,
    ) -> Result<()> {
        let change_info = change.info();
        let bytes = StoredValue::from(storage::ChangeInfo::from(&change_info))
            .serialize()
            .wrap_err("failed to serialize change info")?;
        self.put_raw(keys::change(upgrade_name, &change.name()), bytes);
        Ok(())
    }
}

impl<T: StateWrite> StateWriteExt for T {}

#[cfg(test)]
mod tests {
    use cnidarium::StateDelta;

    use super::*;

    const UPGRADE_1: UpgradeName = UpgradeName::new("up one");

    #[derive(borsh::BorshSerialize)]
    struct TestChange;

    impl Change for TestChange {
        fn name(&self) -> ChangeName {
            ChangeName::from("test_change")
        }

        fn activation_height(&self) -> u64 {
            10
        }

        fn app_version(&self) -> u64 {
            2
        }
    }

    #[tokio::test]
    async fn change_info_roundtrip() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        assert!(state
            .get_upgrade_change_info(&UPGRADE_1, &TestChange.name())
            .await
            .unwrap()
            .is_none());

        state
            .put_upgrade_change_info(&UPGRADE_1, &TestChange)
            .unwrap();

        assert_eq!(
            state
                .get_upgrade_change_info(&UPGRADE_1, &TestChange.name())
                .await
                .unwrap(),
            Some(TestChange.info()),
        );
    }
}
