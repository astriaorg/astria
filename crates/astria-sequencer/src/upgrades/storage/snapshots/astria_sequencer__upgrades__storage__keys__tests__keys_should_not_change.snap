---
source: crates/astria-sequencer/src/upgrades/storage/keys.rs
expression: "change(&UpgradeName::new(\"aspen\"), &ChangeName::new(\"change_1\"))"
---
upgrades/aspen/change_1
